language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToTransformFunctionCodeAction.java

```diff
@@ -0,0 +1,275 @@
+                valueExprNode = specificFieldNode.valueExpr().orElseThrow();
+            } else {
+                specificFieldNode = (SpecificFieldNode) matchedCodeActionNode.parent();
+                valueExprNode = (ExpressionNode) matchedCodeActionNode;
+            }
+
+            semanticModel = context.currentSemanticModel().orElseThrow();
+            TypeSymbol matchedTypeSymbol = posDetails.matchedTopLevelTypeSymbol();
+            Symbol inputSymbol = matchedTypeSymbol == null ?
+                    semanticModel.symbol(context.nodeAtRange()).orElseThrow() : matchedTypeSymbol;
+            inputType = getInputRecord(context, getTypeSymbol(inputSymbol).orElseThrow()).orElseThrow();
+
+            Symbol outputRecordSymbol = semanticModel.symbol(specificFieldNode.fieldName()).orElseThrow();
+            outputRecord = getOutputRecord(context,
+                    getTypeSymbol(outputRecordSymbol).orElseThrow()).orElseThrow();
+
+            enclosingNode = CommonUtil.getMatchingNode(specificFieldNode,
+                    node -> node.parent().kind() == SyntaxKind.MODULE_PART).orElseThrow();
+            currentDocument = context.currentDocument().orElseThrow();
+        } catch (RuntimeException e) {
+            return Collections.emptyList();
+        }
+
+        // Extracting the line position and the range
+        LinePosition functionEndLine = enclosingNode.lineRange().endLine();
+        Range extractedRange = PositionUtil.toRange(valueExprNode.lineRange());
+
+        // Generating the transformation function and the function call
+        List<Symbol> visibleSymbols = semanticModel.visibleSymbols(currentDocument, functionEndLine);
+        String functionName = FunctionGenerator.generateFunctionName(EXTRACTED_PREFIX, visibleSymbols);
+        String extractedFunction = getFunction(functionName, outputRecord, inputType, valueExprNode, visibleSymbols);
+        String functionCall = functionName + CommonKeys.OPEN_PARENTHESES_KEY +
+                valueExprNode.toSourceCode().stripTrailing() +
+                CommonKeys.CLOSE_PARENTHESES_KEY;
+
+        // Generating the text edits and the code action
+        List<TextEdit> textEdits = List.of(
+                new TextEdit(PositionUtil.toRange(functionEndLine), extractedFunction),
+                new TextEdit(extractedRange, functionCall)
+        );
+        CodeAction codeAction = CodeActionUtil.createCodeAction(CommandConstants.EXTRACT_TO_TRANSFORM_FUNCTION,
+                textEdits, context.fileUri(), CodeActionKind.RefactorExtract);
+        CodeActionUtil.addRenamePopup(context, codeAction, CommandConstants.RENAME_COMMAND_TITLE_FOR_FUNCTION,
+                extractedRange.getStart());
+        return Collections.singletonList(codeAction);
+    }
+
+    @Override
+    public List<SyntaxKind> getSyntaxKinds() {
+        return List.of(
+                SyntaxKind.FIELD_ACCESS,
+                SyntaxKind.SPECIFIC_FIELD,
```

Comment:
Don'tbwe need to handle this for `OPTIONAL_FIELD_ACCESS` syntax kind

Reply:
Added a remark regarding this: https://github.com/ballerina-platform/ballerina-lang/pull/42194#pullrequestreview-1891923376:~:text=Optional%20fields%20are%20not%20supported%2C%20as%20we%20need%20a%20mechanism%20on%20how%20we%20should%20handle%20nullable%20types.

Reply:
`getSyntaxKinds()` does not support `SIMPLE_NAME_REFERENCE` because it considers the parent of such nodes (This is a design decision within the LS considering the majority of use cases), As shown in the following test, the CA supports simple name reference nodes.

https://github.com/ballerina-platform/ballerina-lang/blob/01936d3e76ad247bbb2654c58c56591dd55e035c/language-server/modules/langserver-core/src/test/resources/codeaction/extract-to-transform-function/source/extract_to_transform_function12.bal#L31